### DISCUSSION TITLE
Quick Orientation

### DISCUSSION THREAD

**Discussion by johnlindquist** [johnlindquist/kit#131] [General]
For people new to Script Kit, here's a tl;dr

### Simplest possible `arg`:
```js
let input = await arg()
```

Most commonly, you'll pass in a `placeholder` and `choices`

### Using a `placeholder` and `choices`:
```js
let city = await arg("Pick a city", ["Anchorage", "Portland", "Austin"])
```

### Customizing choice values:
```js
let number = await arg("Pick an award", [
  {
    name: "Gold",
    value: 1,
    description: "First place",
  },
  {
    name: "Silver",
    value: 2,
    description: "Second place",
  },
  {
    name: "Bronze",
    value: 3,
    description: "Third place",
  },
])
```

### Choice Functions
Beyond that, a function that returns an Array like above also works.
```js
await arg("Select", ()=> ["one", "two", "three"])
```

If you want to _generate_ choices when you type (instead of the default _filter_ behavior), then pass `input` into the function
```js
let message = await arg("Select", (greeting)=> [`${greeting}, John`, `${greeting}, Mindy`])
```

Generating choices also handles "searching" where you can pass input to apis and return results. These functions can be `async` as well, so many of the helpers such as `await get` to query apis would work inside the function.

## 2. Global 3rd Party Functions

There's a sizeable list of global functions available that alias popular npm packages for common tasks: `get`, `copy`, `replace`, `md`, etc. Check out the [@johnlindquist/globals](https://github.com/johnlindquist/globals/tree/main/src)
 lib to see most of the 3rd party helper functions available. 

# 3. `npm`

The `npm` helper will prompt the user to install a package the first time they run the script:

```js
let sharp = await npm("sharp")
```

> Add code-hinting by using JSDoc like so. We'll work on an official VS Code snippet lib in the future
```js
/** @type typeof import("project-name-generator") */
let generate = await npm("project-name-generator")
```

## 4. Lots more

You can add tabs, a database, previews, drag/drop, and more, but I would say the majority of scripts are a combination of the ideas above.

I'm happy to answer any questions. Happy Scripting! - John

**Comment by Antonio-Laguna** [comment]
How can you get other node native packages? Such as `net` or `path`?

**Reply by Antonio-Laguna** [reply]
I think I've just figured, just `npm` helper works fine.

**Reply by johnlindquist** [reply]
@Antonio-Laguna `path` is global. I figured it would be highly used

**Comment by jessegavin** [comment]
@johnlindquist can you please update the link in "2. Global Helper Functions" ? It currently leads to 404

**Reply by johnlindquist** [reply]
Sorry for missing your comment. 

I updated the post to reflect the [@johnlindquist/globals](https://github.com/johnlindquist/globals/tree/main/src) lib that's now bundled with each script.
